client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -415,9 +415,15 @@ ClientConnection getConnection() throws SegmentSealedException {
     void setupConnection() throws ConnectionFailedException {
         if (state.getConnection() == null) {
             CompletableFuture<ClientConnection> newConnection = controller.getEndpointForSegment(segmentName)
-                .thenCompose((PravegaNodeUri uri) -> {
-                    return connectionFactory.establishConnection(uri, responseProcessor);
-                });
+                    .thenCompose((PravegaNodeUri uri) -> {
+                        return connectionFactory.establishConnection(uri, responseProcessor);
+                    }).whenComplete((connection, ex) -> {
+                        if (ex != null) {
```

Comment:
Yes, I am incorrect here.
* if SSS go down response processor callback will be invoked which will internally invoke state.failConnection
* on a next call since state.getConnection is null we will again invoke cf.establishConnection. And if it fails getAndHandleExceptions will throw an error and hence connection will remain null.